File: ch2/Binary_Search/algorithm.cs
using System;

public class algorithm
{
    public static int binary_search<T>(T[] L, T X) where T: IComparable<T>
    {
        int low = 0;
        int high = L.Length -1;
        while (low<=high) {
            int mid = (high+low)/2;
            if (L[mid].CompareTo(X)==0){
                return mid;
            } else if (L[mid].CompareTo(X)>0) {
                low = mid+1;
            } else if (L[mid].CompareTo(X)<0) {
                high = mid-1;
            }
        }
        return -1;
    }

    public static void Main(string[] args) {
        int[] numbers = {1,4,123,123,54,3423};
        Console.Write(algorithm.binary_search<int>(numbers, 54));
    }
}

File: ch2/Binary_Search/algorithm.go
package main

import "fmt"

func binary_search(List []int, X int) int {
	low := 0
	high := len(List) - 1
	for low <= high {
		mid := (low+high)/2
		if List[mid] == X {
			return mid
		} else if List[mid] > X {
			high = mid - 1
		} else if List[mid] < X {
			low = mid + 1
		}
	}
	return -1
}

func main() {
	List := []int{2,6,9,10,13,16,18,19,23,24,29,31,39,41,46,51,56,62,67,70,74}
	fmt.Println(binary_search(List, 23))
}